helm: adds annotations support to certgen Job and CronJob

eripa opened the pull request:
Some tooling, such as Argo CD reads resource annotations to signal intent and sync policies.

A specific example is that the certgen `Job` as it currently is will result sync out-of-sync statuses as when the bootstrap  certgen Job is cleaned up.

This can be prevented by adding an annotation to the Job, such as:

```yaml
  annotations:
    argocd.argoproj.io/hook: PreSync
    argocd.argoproj.io/hook-delete-policy: HookSucceeded
```

Note: this is an additive feature and is backwards compatible


```release-note
helm: Allow adding annotations to certgen Job and CronJob
```

kaworu reviewed (changes requested):
Hi @eripa and thank you for the PR! Patch LGTM, but please report the Helm values changes to `install/kubernetes/cilium/values.yaml.tmpl`, and then run `make -C install/kubernetes cilium/values.yaml` to update the `values.yaml` file. Then, [this CI failure](https://app.travis-ci.com/github/cilium/cilium/jobs/589451939) needs to be fixed:

> HINT: to fix this, run 'make -C Documentation update-helm-values'

eripa committed:
helm: adds annotations support to certgen

Some tooling, such as Argo CD use annotations to
signal intent and sync policies.

A specific example is that the certgen Job as it
currently is will result sync out-of-sync statuses
as when the bootstrap certgen Job is cleaned up.

This can be prevented by adding an annotation to
the Job, such as:

  annotations:
    argocd.argoproj.io/hook: PreSync
    argocd.argoproj.io/hook-delete-policy: HookSucceeded

Note: this is an additive feature and is backwards
compatible

Signed-off-by: Eric Ripa <eric@ripa.io>

kaworu reviewed (changes requested):
Thanks for the update @eripa ! Helm charts CI [failed with](https://github.com/cilium/cilium/actions/runs/3551360606/jobs/5986223092)

> please run 'make -C install/kubernetes' and submit your changes

After that the patch should be ready!

kaworu reviewed (approved):
:rocket:

tommyp1ckles commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -758,6 +758,10 @@ certgen:
   # -- Labels to be added to hubble-certgen pods
   podLabels: {}
+  # -- Annotations to be added to the hubble-certgen initial Job and CronJob
+  annotations:
```
seems like we generally do this with either something like:

```yaml
cilium:
    annotations: {}
```
or

```yaml
podAnnotations: {}
```

Perhaps it would be more consistent to do something like:

```
...
    job:
        annotations: {}
    cronJob:
        annotations: {}
```

for consistency?

eripa replied:
Thanks for the feedback @tommyp1ckles 

I actually made it like this in the hope that it would be consistent, or at least similar look & feel.

modeled after `podAnnotations: {}` and `podLabels: {}` it would most naturally be

```yaml
jobAnnotations: {}
cronJobAnnotations: {}
```

however.. adding more and more single-level keys like this isn't a great solution. So I opted for nesting it under one annotations keys in order to make it more future proof. I.e. it could eventually be:

```yaml
annotations:
    all: {}
    job: {}
    cronjob: {}
    pod: {}
labels: 
    all: {}
    pod: {} # theoretically podLabels could be moved here 
    job: {}
    cronJob: {}
```

this makes the object easily extensible, whilst the inverse approach (`.job.annotation` etc.). could result in several single-item objects.

With that said, I'm fine with any approach. And realistically this certgen component isn't going to grow much. 😊 

Let me know what you think.

tommyp1ckles replied:
makes sense to me, agreed on the single level keys not being the best solution.
